### ISSUE TITLE
heatmap-paired unable to find some features

### ISSUE THREAD

**Issue by mestaki** [biocore/mmvec#141]
Hey @mortonjt and others (?),
Finally getting around to play with this mythical tool.

I was able to run standalone mmvec (mmvec 1.0.5)  and then import the results into q2-mmvec (q2-2020.6, mmvec-V 1.0.4) for further analysis of my own data.
I was mostly successful in following all of the steps in the [tutorial](https://github.com/biocore/mmvec) until I reached `mmvec heatmap-paired` step: 

```
!qiime mmvec paired-heatmap \
  --i-ranks conditionals.qza \
  --i-microbes-table filtered_otus_nt.qza \
  --i-metabolites-table filtered_lcms_nt.qza \
  --m-microbe-metadata-file ../microbes/gtdb-bespoke-taxonomy.tsv \
  --m-microbe-metadata-column Taxon \
  --p-features TACGTAGGTCCCGAGCGTTATCCGGATTTATTGGGCGTAAAGCGAGCGCAGGCGGTTAGATAAGTCTGAAGTTAAAGGCTGTGGCTTAACCATAGTACGCTTTGGAAACTGTTTAACTTGAGTGCAGAAGGGGAGAGTGGAATTCCATGT \
  --p-features TACGGAGGGTGCAAGCGTTAATCGGAATTACTGGGCGTAAAGCGCACGCAGGCGGTCTGTCAAGTCGGATGTGAAATCCCCGGGCTCAACCTGGGAACTGCATTCGAAACTGGCAGGCTAGAGTCTTGTAGAGGGGGGTAGAATTCCAGG \
  --p-features TACGGAGGGTGCAAGCGTTAATCGGAATTACTGGGCGTAAAGCGCACGCAGGCGGTTTGTTAAGTCAGATGTGAAATCCCCGGGCTCAACCTGGGAACTGCATCTGATACTGGCAAGCTTGAGTCTCGTAGAGGGGGGTAGAATTCCAGG \
  --p-features TACGGAGGGTGCAAGCGTTAATCGGAATTACTGGGCGTAAAGCGCACGCAGGCGGTCTGTCAAGTCGGATGTGAAATCCCCGGGCTCAACCTGGGAACTGCATCCGAAACTGGCAGGCTAGAGTCTTGTAGAGGGGGGTAGAATTCCAGG \
  --p-top-k-microbes 0 \
  --p-normalize rel_row \
  --p-top-k-metabolites 100 \
  --p-level 7 \
  --o-visualization paired-heatmap-top2.qzv
```

> Traceback (most recent call last):
>   File "/home/mestaki/miniconda3/envs/qiime2-2020.6/lib/python3.6/site-packages/q2cli/commands.py", line 329, in __call__
>     results = action(**arguments)
>   File "<decorator-gen-540>", line 2, in paired_heatmap
>   File "/home/mestaki/miniconda3/envs/qiime2-2020.6/lib/python3.6/site-packages/qiime2/sdk/action.py", line 245, in bound_callable
>     output_types, provenance)
>   File "/home/mestaki/miniconda3/envs/qiime2-2020.6/lib/python3.6/site-packages/qiime2/sdk/action.py", line 452, in _callable_executor_
>     ret_val = self._callable(output_dir=temp_dir, **view_args)
>   File "/home/mestaki/miniconda3/envs/qiime2-2020.6/lib/python3.6/site-packages/mmvec/q2/_visualizers.py", line 72, in paired_heatmap
>     color_palette)
>   File "/home/mestaki/miniconda3/envs/qiime2-2020.6/lib/python3.6/site-packages/mmvec/heatmap.py", line 202, in paired_heatmaps
>     select_metabolites = metabolites_table[top_metabolites]
>   File "/home/mestaki/miniconda3/envs/qiime2-2020.6/lib/python3.6/site-packages/pandas/core/frame.py", line 3001, in __getitem__
>     indexer = self.loc._convert_to_indexer(key, axis=1, raise_missing=True)
>   File "/home/mestaki/miniconda3/envs/qiime2-2020.6/lib/python3.6/site-packages/pandas/core/indexing.py", line 1285, in _convert_to_indexer
>     return self._get_listlike_indexer(obj, axis, **kwargs)[1]
>   File "/home/mestaki/miniconda3/envs/qiime2-2020.6/lib/python3.6/site-packages/pandas/core/indexing.py", line 1092, in _get_listlike_indexer
>     keyarr, indexer, o._get_axis_number(axis), raise_missing=raise_missing
>   File "/home/mestaki/miniconda3/envs/qiime2-2020.6/lib/python3.6/site-packages/pandas/core/indexing.py", line 1177, in _validate_read_indexer
>     key=key, axis=self.obj._get_axis_name(axis)
> KeyError: "None of [Int64Index([6102, 2467, 6103, 6104, 6107, 6114, 2513, 6087, 6118, 4831,\n            ...\n            3520, 9554, 6686, 6691, 2873, 2805, 1234, 4882, 3794, 2934],\n           dtype='int64', length=263)] are in the [columns]"
> 
> Debug info has been saved to /tmp/qiime2-q2cli-err-m2xxease.log

This error message is similar to [133](https://github.com/biocore/mmvec/issues/133), and I have checked all the potential causes suggested there to no avail. The ultimate solution that fixed the user's issue there does not apply in my case because my sample names and metabolite featureids do not match to begin with, this leads me to believe something else fishy is going. 

My sample names are: `2136.000006328`, `2136.000006340`...
My metabolite featureids are numeric: `1`, `2`...`9780`

I have manually checked the missing features in the error log in my filtered_lcms_nt.qza (converted to .qzv) and they all do exist there.
One thing worth mentioning, when I change or even remove all the ASVs in `--p-features`, the list of missing features that the error log complains about  changes. So far the one thing that was common in the other issue is that featureid are integers. I'm not sure if the other user there changed their featureids to something non-numeric as well or not, but could these by causing some issues?

I don't think there are any issues with the orientation of my files/bioms etc. as otherwise I wouldn't have been able to get this far. I've also checked with some of the online tutorial material and they match (ex. CF) those.

Any help would be appreciated!

**Comment by mortonjt** [comment]
Hi @mestaki , I'm wondering if there is a similar error to #133. How did you create the biom table for the metabolites?
Perhaps the easiest workaround is to explicitly convert all of the metabolite names to strings (i.e. `X1`, `X2`, ...)

**Comment by mestaki** [comment]
Thanks @mortonjt , the microbe biom table is from Qiita. The metabolite dataframe was curated in R, exported, then transformed with **biom**.

Actually, just remembered, both of these were then modified with python then exported again, but the codes were based exactly on your `Figure4-cystic-fibrosis.ipynb` notebook. 
Where the final step:

```
from biom import Table
from biom.util import biom_open
microbe_table_t0 = Table(microbes_t0.values.T, microbes_t0.columns, microbes_t0.index)
metabolite_table_t0 = Table(metabolites_t0.values.T, metabolites_t0.columns, metabolites_t0.index)
#sanity check
microbe_table_t0, metabolite_table_t0

(142 x 70 <class 'biom.table.Table'> with 2893 nonzero entries (29% dense),
 9780 x 70 <class 'biom.table.Table'> with 656040 nonzero entries (95% dense))
```
and exported as biom again

```
with biom_open('../mmvec_t0/filtered_otus_nt.biom', 'w') as f:
    microbe_table_t0.to_hdf5(f, 'matched')

with biom_open('../mmvec_t0/filtered_lcms_nt.biom', 'w') as f:
    metabolite_table_t0.to_hdf5(f, 'matched')
```

I'll give the renaming a try. I should be able to just edit the names in the mmvec-paired outputs right, i.e. ranks and conditions texts and my metadata files? Nothing upstream that would interfere with that, that would require me re-running the whole thing?

**Comment by mortonjt** [comment]
Renaming the metabolites should fix the problem. If you explicitly convert
the metabolites to strings, that'll be another solution. You shouldn't have
to rerun mmvec under any of those scenarios

metabolite_table_t0 = Table(metabolites_t0.values.T, list(map(str,
metabolites_t0.columns)) , metabolites_t0.index)
#sanity check
microbe_table_t0, metabolite_table_t0

**Comment by mestaki** [comment]
Hey @mortonjt ,
So saving the metabolite numeric features explicitly as string by map(str) did not work, I ended up with the same error message. Not sure if that property is lost at the biom export or qiime2 import step but yeah...didn't work.
On the other hand I went and added an "X" prefix to all those feature-ids in the metabolite.biom as well as the features in the rankings file (mmvec output) and then it worked! 
So I guess for the time being I would recommend not using numeric values for feature-ids.
